### PR TITLE
feat: Define TOTP recipe plugin interface

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Define TOTP recipe plugin interface

## Related issues
- https://github.com/supertokens/supertokens-core/issues/554

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] Changes to the version if needed
   - In `build.gradle`
- [ ] Had installed and ran the pre-commit hook
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import io.supertokens.pluginInterface.totp.TOTPDevice;
+import io.supertokens.pluginInterface.totp.exception.DeviceAlreadyExistsException;
+import io.supertokens.pluginInterface.totp.exception.InvalidCodeException;
+import io.supertokens.pluginInterface.totp.exception.LimitReachedException;
+import io.supertokens.pluginInterface.totp.exception.NotEnabledException;
+import io.supertokens.pluginInterface.totp.exception.UnknownDeviceException;
+
+public interface TOTPStorage extends AuthRecipeStorage {
+    // Create a new device:
+    void createDevice(String userId, String deviceName, String secret, int skew, int period)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:16]:
```suggestion
    void createDevice(TotpDevice device)
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package io.supertokens.pluginInterface.totp;
+
+public class TOTPDevice {
+    public final String deviceName;
+    public final String userId;
+    public final String secretKey;
+    public final int period;
+    public final int skew;
+    public final Boolean verified;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java:9]:
```suggestion
    public final boolean verified;
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package io.supertokens.pluginInterface.totp;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java:1]: Add totp to RECIPE_ID.java

File: src/main/java/io/supertokens/pluginInterface/totp/exception/DeviceAlreadyExistsException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package io.supertokens.pluginInterface.totp.exception;
+
+import io.supertokens.pluginInterface.emailpassword.exceptions.EmailPasswordException;
+
+// Why extending EmailPasswordException everywhere ?
+public class DeviceAlreadyExistsException extends EmailPasswordException {
+    private static final long serialVersionUID = 6848053563771647272L; // How's this decided ?
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/exception/DeviceAlreadyExistsException.java:7]: remove all this

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    // Delete the device with the given name:
+    void deleteDevice(String userId, String deviceName) throws StorageQueryException, NotEnabledException;
+
+    // update device name:
+    void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+            throws StorageQueryException, NotEnabledException, DeviceAlreadyExistsException, UnknownDeviceException;
+
+    // Get the devices
+    TOTPDevice[] getDevices(String userId, @Nullable String deviceName)
+            throws StorageQueryException, NotEnabledException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:36]:
```suggestion
    TOTPDevice[] getDevices(String userId)
            throws StorageQueryException, NotEnabledException;
```

File: src/main/java/io/supertokens/pluginInterface/totp/exception/NotEnabledException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package io.supertokens.pluginInterface.totp.exception;
+
+import io.supertokens.pluginInterface.emailpassword.exceptions.EmailPasswordException;
+
+public class NotEnabledException extends EmailPasswordException {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/exception/NotEnabledException.java:5]:
```suggestion
public class TotpNotEnabledException extends EmailPasswordException {
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    // Get all the device names for the given user:
+    String[] getDeviceNames(String userId) throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:39]:
```suggestion
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    // Create a new device:
+    void createDevice(String userId, String deviceName, String secret, int skew, int period)
+            throws StorageQueryException, DeviceAlreadyExistsException;
+
+    // Verify the TOTP code:
+    void verifyCode(String userId, String deviceName, String code)
+            throws StorageQueryException, InvalidCodeException, NotEnabledException, LimitReachedException;
+
+    // Verify the device with the given name:
+    void verifyDevice(String userId, String deviceName)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:24]:
```suggestion
    void markDeviceAsVerified(String userId, String deviceName)
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            throws StorageQueryException, InvalidCodeException, NotEnabledException, UnknownDeviceException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:25]:
```suggestion
            throws StorageQueryException, NotEnabledException, UnknownDeviceException;
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import io.supertokens.pluginInterface.totp.exception.UnknownDeviceException;
+
+public interface TOTPStorage extends AuthRecipeStorage {
+    // Create a new device:
+    void createDevice(String userId, String deviceName, String secret, int skew, int period)
+            throws StorageQueryException, DeviceAlreadyExistsException;
+
+    // Verify the TOTP code:
+    void verifyCode(String userId, String deviceName, String code)
+            throws StorageQueryException, InvalidCodeException, NotEnabledException, LimitReachedException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:21]:
```suggestion
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    // Verify the device with the given name:
+    void verifyDevice(String userId, String deviceName)
+            throws StorageQueryException, InvalidCodeException, NotEnabledException, UnknownDeviceException;
+
+    // Delete the device with the given name:
+    void deleteDevice(String userId, String deviceName) throws StorageQueryException, NotEnabledException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:28]:
```suggestion
    void deleteDevice(String userId, String deviceName) throws StorageQueryException, NotEnabledException, UnknownDeviceException;
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+            throws StorageQueryException, NotEnabledException, DeviceAlreadyExistsException, UnknownDeviceException;
+
+    // Get the devices
+    TOTPDevice[] getDevices(String userId, @Nullable String deviceName)
+            throws StorageQueryException, NotEnabledException;
+
+    // Get all the device names for the given user:
+    String[] getDeviceNames(String userId) throws StorageQueryException;
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:40]: Need functions for:
- inserting into totp_used_codes
- getting info from totp_used_codes based on userId
- remove expired codes from totp_used_codes

File: src/main/java/io/supertokens/pluginInterface/totp/exception/LimitReachedException.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package io.supertokens.pluginInterface.totp.exception;
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/exception/LimitReachedException.java:2]: remove this exception

File: src/main/java/io/supertokens/pluginInterface/totp/exception/InvalidCodeException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package io.supertokens.pluginInterface.totp.exception;
+
+import io.supertokens.pluginInterface.emailpassword.exceptions.EmailPasswordException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/exception/InvalidCodeException.java:3]: remove this

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package io.supertokens.pluginInterface.totp;
+
+public class TOTPUsedCode {
+    public final String userId;
+    public final String code;
+    public final Boolean isValidCode;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java:6]:
```suggestion
    public final boolean isValidCode;
```

Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java:6]:
```suggestion
    public final boolean isValidCode;
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java (hunk 3 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java:1]: done.

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java (hunk 2 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPDevice.java:9]: done.

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 10 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:21]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 1 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:16]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 8 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:24]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 9 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:25]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 11 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:28]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 5 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:36]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java (hunk 15 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java:6]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 7 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:39]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java (hunk 15 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPUsedCode.java:6]: done

File: src/main/java/io/supertokens/pluginInterface/totp/exception/NotEnabledException.java (hunk 6 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/exception/NotEnabledException.java:5]: done.

File: src/main/java/io/supertokens/pluginInterface/totp/exception/LimitReachedException.java (hunk 13 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/exception/LimitReachedException.java:2]: done

File: src/main/java/io/supertokens/pluginInterface/totp/exception/InvalidCodeException.java (hunk 14 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/exception/InvalidCodeException.java:3]: done.

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 12 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java]: done

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 16]
```diff
@@ -4,38 +4,41 @@
...
+                        throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException;
+
+        // update device name:
+        void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+                        throws StorageQueryException, TotpNotEnabledException, DeviceAlreadyExistsException,
+                        UnknownDeviceException;
+
+        // Get the devices
+        TOTPDevice[] getDevices(String userId)
+                        throws StorageQueryException, TotpNotEnabledException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:31]:
```suggestion
        TOTPDevice[] getDevices(String userId)
                        throws StorageQueryException;
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 17]
```diff
@@ -4,38 +4,41 @@
...
+
+        // Insert a used TOTP code:
+        void insertUsedCode(TOTPUsedCode code)
+                        throws StorageQueryException;
+
+        // Get totp used code based on userId:
+        TOTPUsedCode[] getUsedCodes(String userId)
+                        throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:39]:
```suggestion
        void insertUsedCode(TOTPUsedCode code)
                        throws StorageQueryException, TotpNotEnabledException;

        // Get totp used code based on userId:
        TOTPUsedCode[] getUsedCodes(String userId)
                        throws StorageQueryException, TotpNotEnabledException;
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 18]
```diff
@@ -4,38 +4,41 @@
...
+        void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+                        throws StorageQueryException, TotpNotEnabledException, DeviceAlreadyExistsException,
+                        UnknownDeviceException;
+
+        // Get the devices
+        TOTPDevice[] getDevices(String userId)
+                        throws StorageQueryException, TotpNotEnabledException;
+
+        // Insert a used TOTP code:
+        void insertUsedCode(TOTPUsedCode code)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:34]:
```suggestion
        boolean insertUsedCode(TOTPUsedCode code)
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/exception/TotpNotEnabledException.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package io.supertokens.pluginInterface.totp.exception;
+
+import io.supertokens.pluginInterface.emailpassword.exceptions.EmailPasswordException;
+
+public class TotpNotEnabledException extends EmailPasswordException {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/exception/TotpNotEnabledException.java:5]: These should not extend EmailPasswordException. Instead, they should all just extend Exception.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/pluginInterface/totp/exception/TotpNotEnabledException.java:5]: This bug is also in other exceptions of this interface.. i know. But at least fix these totp exception types

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+import io.supertokens.pluginInterface.totp.exception.UnknownDeviceException;
+
+public interface TOTPStorage extends AuthRecipeStorage {
+        // Create a new device:
+        void createDevice(TOTPDevice device)
+                        throws StorageQueryException, DeviceAlreadyExistsException;
+
+        // Verify the device with the given name:
+        void markDeviceAsVerified(String userId, String deviceName)
+                        throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:18]: i don't see the usefulness of throwing `TotpNotEnabledException` from these functions ( markDeviceAsVerified, deleteDevice, updateDeviceName)- cause the primary key in the tables is (user_id, device_name). So failure to update those rows will always just result in `UnknownDeviceException` exception.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 21]
```diff
@@ -9,36 +9,43 @@
 import io.supertokens.pluginInterface.totp.exception.UnknownDeviceException;
 
 public interface TOTPStorage extends AuthRecipeStorage {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:11]:
```suggestion
public interface TOTPStorage extends NonAuthRecipeStorage {
```

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 22]
```diff
@@ -9,36 +9,43 @@
...
         TOTPUsedCode[] getNonExpiredUsedCodes(String userId)
                         throws StorageQueryException;
 
-        // Remove expired codes from totp used codes:
+        /** Remove expired codes from totp used codes for all users: */
         void removeExpiredCodes()
                         throws StorageQueryException;
+
+        /** Delete all data associated with the user: */
+        void deleteAllDataForUser(String userId)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:49]:
```suggestion
        void deleteAllTOTPRelatedDataForUser(String userId)
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java [hunk 23]
```diff
@@ -35,17 +31,17 @@ void insertUsedCode(TOTPUsedCode code)
...
         /** Remove expired codes from totp used codes for all users: */
-        void removeExpiredCodes()
+        int removeExpiredCodes()
                         throws StorageQueryException;
 
-        /** Delete all data associated with the user: */
-        void deleteAllDataForUser(String userId)
+        /** Delete all totp data associated with the user: */
+        void deleteAllTotpDataForUser(String userId)
                         throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java:46]: this should not be required anymore

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 23 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java]: Done.

File: src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java (hunk 22 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/totp/TOTPStorage.java]: Removed now.

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

File: src/main/java/io/supertokens/pluginInterface/totp/sqlStorage/TOTPSQLStorage.java [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+import io.supertokens.pluginInterface.totp.TOTPUsedCode;
+import io.supertokens.pluginInterface.totp.exception.TotpNotEnabledException;
+import io.supertokens.pluginInterface.totp.exception.UsedCodeAlreadyExistsException;
+
+public interface TOTPSQLStorage extends TOTPStorage, SQLStorage {
+    public int deleteDevice_Transaction(TransactionConnection con, String userId, String deviceName)
+            throws StorageQueryException;
+
+    public int getDevicesCount_Transaction(TransactionConnection con, String userId)
+            throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/totp/sqlStorage/TOTPSQLStorage.java:16]: remove this and add getDevices_Transaction(TransactionConnection con, String userId)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**